_posts/2018-05-08-pulp3-moving-to-rq.md

```diff
@@ -0,0 +1,162 @@
+---
+title: Pulp 3 Moving to RQ for Tasking
+author: Brian Bouterse
+tags:
+  - 3.0
+---
+
+Pulp adopted the multi-processing tasking system in Pulp 2.4. Since then, users have benefited from
+the throughput and availability it provides, but core developers regularly receive bug reports with
+symptoms like:
+
+* Pulp stops processing tasks entirely
+* A single task never starts
+
+Ensuring the reliability of Pulp's tasking system and therefore Pulp itself is a paramount concern.
+After careful consideration, the maintainers of Pulp's tasking system believe the best way to
+resolve these reliability issues is by porting the Pulp 3.0 tasking system from using
+[Celery](http://celery.readthedocs.io/en/latest/) to using [RQ](http://python-rq.org/docs/).
+
+This change will be made on May 15, 2018, and will be released with pulpcore Beta 4 on May 16, 2018.
+
+Pulp 2.y is being left as-is because the change is backwards incompatible, so we can't introduce
+them due to [Semantic Versioning](http://semver.org/).
+
+This post is divided into three sections: Reasoning, User Changes, Plugin Writer Changes
+
+## Reasoning
+
+Four primary reasons stood out motivating this decision.
+
+1. Users are experiencing task-halting issues which users and developers alike cannot reproduce. We
+   have resolved some of these, but there are some that are unresolved. New ones are reported
+   periodically, so there is no end in sight.
+
+2. The more time core developers spend fixing Celery, the less time we are improving Pulp and its
+   plugins.
+
+3. It's epically complicated for little benefit. The code of Celery itself is spread across Celery,
+   Kombu, pyamqp, billiard, qpid.messaging, and other libraries. With RQ it's just a lot simpler.
+
+4. RQ fulfills 100% of the feature needs of Pulp with no gaps.
+
+More discussion that went into this decision can be found in the "Port Pulp3 to use RQ" thread on
+the pulp-dev mailing list archives for
+[March](https://www.redhat.com/archives/pulp-dev/2018-March/thread.html),
+[April](https://www.redhat.com/archives/pulp-dev/2018-April/thread.html), and
+[May](https://www.redhat.com/archives/pulp-dev/2018-May/thread.html).
+
+
+## User Changes
+
+### Redis replaces RabbitMQ/Qpid
+
+RQ uses Redis as its backend, so with this change, Redis replaces RabbitMQ or Qpid in the Pulp
+architecture. Y
+
+### Run workers differently
+
```

Comment:
The out of place ```instead of doing X:``` was removed but I feel like instead of removing it entirely it should go here.

Reply:
fixed in next push.